Fix exception which is thrown on search results page

wojtekn opened the pull request:
### Description

PR fixes issue which results with throwing exception on search page when phrase contains slash (/).

### Related Pull Requests
n/a

### Fixed Issues (if relevant)

1. Fixes magento/magento2#25886
2. Fixes magento/magento2#25110
3. Fixes magento/magento2#28286

### Manual testing scenarios

1. Install clean Magento
2. Ensure that Elasticsearch is enabled as search backend
3. Ensure the synonym exists in SEO & Search -> Search Synonyms:
foo,bar"foobar",
2. Search for "Foo/Bar" in frontend

Without fix, the exception is thrown. PR ensures that generated phrases have preg modifiers escaped so exception won't be thrown.

### Questions or comments

Fix was made inside private method. I'm not sure if any tests should be changed for that? Or regression test should be added?

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

wojtekn committed:
Fix exception which is thrown on search results page

wojtekn committed:
Modify unit test to cover for buggy use case

wojtekn commented:
I reviewed testGetSynonymsForPhrase and it was not including slash, so the buggy use case was not caught by that test.

I was able to make small modification to the test to cover for the change. Test fails when fix is not in place, it goes through when fix is applied.

lbajsarowicz commented:
@magento run all tests

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-8303](https://jira.corp.magento.com/browse/ENGCOM-8303) has been created to process this Pull Request
:eight_spoked_asterisk: @sidolov, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Alfa commented:
@magento give me 2.4-develop instance

engcom-Alfa reviewed (changes requested):
Hi @wojtekn .

Could you fix failing tests?
Thanks!

wojtekn committed:
Add missing line break before docblock tags

wojtekn commented:
@magento run Static Tests

wojtekn commented:
@magento run Integration Tests

wojtekn committed:
Change implementation to ensure one of current use cases is not broken. Add case with slash in regression integration test.

wojtekn commented:
@magento run all tests

wojtekn committed:
Fix incorrect line indent

wojtekn commented:
@magento run all tests

wojtekn commented:
@engcom-Alfa thanks, I changed implementation as my initial fix was breaking one use case. I also fixed static test and added regression integration test for use case which my PR is fixing.

Now all tests besides of Functional Tests EE  went through. I checked Allure report but I can't see anything there.

Yesterday I saw that on another run other 5 or 6 tests failed due to errors related to inventory modules. Any suggestions?

sivaschenko commented:
@magento run Functional Tests EE

sivaschenko commented:
Thanks for updates @wojtekn ! Let's give Functional Tests EE one more chance)

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-8303](https://jira.corp.magento.com/browse/ENGCOM-8303) has been created to process this Pull Request

lbajsarowicz commented:
@magento run Functional Tests EE

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-8303](https://jira.corp.magento.com/browse/ENGCOM-8303) has been created to process this Pull Request

engcom-Alfa reviewed (changes requested):
Hi @wojtekn .

Unfortunately, we are not able to reproduce the issue on fresh 2.4-develop. It looks like the problem was already fixed.

**Manual testing scenario:**

1. Install clean Magento;
2. Create simple product "Foo/Bar"
3. Ensure that Elasticsearch is enabled as search backend
4. Ensure the synonym exists in SEO & Search -> Search Synonyms:
foo,bar"foobar",
5. Search for "Foo/Bar" in frontend

**Actual Result:** :heavy_check_mark: Find that particular product without errors.

![2020-10-08_11-29](https://user-images.githubusercontent.com/51679138/95434160-9b77ef80-0959-11eb-878e-9b4a7cc969dc.png)

@wojtekn Could you take a look?

Thanks!

wojtekn commented:
@engcom-Alfa I've checked it, it looks that escaping was added in \Magento\Elasticsearch\SearchAdapter\Query\ValueTransformer\TextTransformer::escape indeed. I think I was reproducing it on 2.4 on my local, I'm not sure what happened.

However, issue will still occur if phrase which contains not escaped slash will be passed to public 
\Magento\Search\Model\SynonymAnalyzer::getSynonymsForPhrase method. I'm not sure if Search module should rely on Elasticsearch module to work properly.

engcom-Alfa commented:
@wojtekn yes, the issue occurs if a phrase which contains not escaped slash will be passed to public
\Magento\Search\Model\SynonymAnalyzer::**getSynonymsForPhrase** method

**Manual testing scenario:**
1. Ensure that Elasticsearch is enabled as search backend
2. Ensure the synonym exists in SEO & Search -> Search Synonyms:
foo,bar"foobar",
3. Add phrase ('**Foo/Bar**') which contains not escaped slash to public
\Magento\Search\Model\SynonymAnalyzer::**getSynonymsForPhrase** method

**Before:** :heavy_multiplication_x: Search results in an error.

![Screenshot from 2020-10-09 12-20-17](https://user-images.githubusercontent.com/51679138/95571208-bd907100-0a30-11eb-83b5-ccc4438665be.png)

**After:** :heavy_check_mark: Find that product without errors.

![2020-10-09_12-32](https://user-images.githubusercontent.com/51679138/95571276-d6992200-0a30-11eb-8910-7172043c8dd4.png)

@sivaschenko @lbajsarowicz @sidolov Could you take a look?

Thanks!

sivaschenko reviewed (approved):
I think we should deliver this fix

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-8303](https://jira.corp.magento.com/browse/ENGCOM-8303) has been created to process this Pull Request

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-8303](https://jira.corp.magento.com/browse/ENGCOM-8303) has been created to process this Pull Request

JonathanHelvey commented:
Hi @wojtekn, We implemented Elastic Search 7.0 on our Magneto 2.3.5 pt-2 staging instance. We added your fix below to file `vendor/magento/module-search/Model/SynonymAnalyzer.php.`

```
            $safeRegexWords = array_map(function ($word) {
                return preg_quote($word, '/');
            }, $words);
```

This fixed all slashes issues, however, we still have an issue when searching only this "Ccell carts 1/2 gm." When searching this on frontend we get thrown this error  below and not sure how to fix. 

We Ensured the synonym exists in SEO & Search -> Search Synonyms:
cartomizer,ccell,extract tank,cart

```
1 exception(s):
Exception #0 (Exception): Warning: preg_match(): Unknown modifier '2' in /var/www/vhosts/upperlimitsinc.com/stagingm2/vendor/magento/module-search/Model/SynonymAnalyzer.php on line 96
```

 On my Localhost the fixed worked with no error thrown, however, on our staging server it now throws this error. 
 
 
Any help would be much appreciated.
